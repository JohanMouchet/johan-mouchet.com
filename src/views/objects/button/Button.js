// @flow

import * as React from "react";
import cx from "classnames";
import "./Button.scss";

type ButtonProps = {
  type: "anchor" | "button" | "submit" | "reset",
  url?: string,
  variant?: "primary" | "secondary",
  size?: "sm" | "lg",
  block?: boolean | "@sm" | "@md",
  className?: string | Array<string> | Object,
  attributes?: { [key: string]: any },
  children?: Node,
};

const Button = ({
  type = "button",
  url,
  variant,
  size,
  block,
  className,
  attributes,
  children,
}: ButtonProps) => {
  const classNames = cx(
    "o-button",
    size && `o-button--${size}`,
    variant && `o-button--${variant}`,
    typeof block === "string"
      ? `o-button--block${"-" + block}`
      : block === true && "o-button--block",
    className
  );

  return type === "anchor" ? (
    <a className={classNames} href={url} {...attributes}>
      {children}
    </a>
  ) : (
    <button className={classNames} type={type} {...attributes}>
      {children}
    </button>
  );
};

export { Button };

type ButtonSocialProps = {
  url: string,
  icon: string,
  counter?: number,
  variant?:
    | "primary"
    | "secondary"
    | "social-twitter"
    | "social-facebook"
    | "social-googleplus"
    | "social-linkedin"
    | "social-pinterest",
  size?: "sm" | "lg",
  block?: boolean | "sm" | "md",
  className?: string | Array<string> | Object,
  attributes?: { [key: string]: any },
  children?: Node,
};

const ButtonSocial = ({
  url,
  icon,
  counter,
  variant,
  size,
  block,
  className,
  attributes,
  children,
}: ButtonSocialProps) => {
  const classNames = cx(
    "o-button",
    size && `o-button--${size}`,
    variant && `o-button--${variant}`,
    typeof block === "string"
      ? `o-button--block${"-" + block}`
      : block === true && "o-button--block",
    "o-button--social",
    className
  );

  return (
    <a className={classNames} href={url} {...attributes}>
      <span className="o-button__icon">{icon}</span>
      <span className="o-button__text">{children}</span>
      {counter && <span className="o-button__counter">{counter}</span>}
    </a>
  );
};

export { ButtonSocial };

type ButtonSocialBadgeProps = {
  url: string,
  icon: string,
  variant?:
    | "primary"
    | "secondary"
    | "social-twitter"
    | "social-facebook"
    | "social-googleplus"
    | "social-linkedin"
    | "social-pinterest",
  className?: string | Array<string> | Object,
  attributes?: { [key: string]: any },
};

const ButtonSocialBadge = ({
  url,
  icon,
  variant,
  className,
  attributes,
}: ButtonSocialBadgeProps) => {
  const classNames = cx(
    "o-button",
    variant && `o-button--${variant}`,
    "o-button--social",
    "o-button--social-badge",
    className
  );

  return (
    <a
      className={classNames}
      href={url}
      target="_blank"
      rel="noopener noreferrer"
      {...attributes}
    >
      <span className="o-button__icon">{icon}</span>
    </a>
  );
};

export { ButtonSocialBadge };
