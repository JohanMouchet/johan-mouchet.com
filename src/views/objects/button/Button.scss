@import "assets/styles/settings/_settings";
@import "assets/styles/tools/_tools";

.o-button {
  $root: &;
  $height: t-vertical-rhythm(1.75) !default;
  $border-width: 2px !default;
  $h-padding: $height / 1.75;

  /**
	 * Btn Padding (private)
	 *
	 * Set button vertical padding to match given button height in order to match inputs height.
	 *
	 * @param {number} [btn-scale-factor = 1] - Increase / decrease factor
	 */
  @mixin btn-padding($btn-scale-factor: 1) {
    @if $height > $s-line-height-base {
      padding: ($height - $s-line-height-base) /
        2 *
        $btn-scale-factor
        $h-padding *
        $btn-scale-factor; // Fallback

      padding: calc(
          #{($height - $s-line-height-base) /
            2 *
            $btn-scale-factor} -
            #{$border-width}
        )
        $h-padding *
        $btn-scale-factor;
    } @else {
      padding: 0 $h-padding;
    }
  }

  @include btn-padding();

  display: inline-block;
  line-height: $s-line-height-base;
  vertical-align: middle;
  border: $border-width solid $s-color-link;
  border-radius: 0;
  background: none;
  color: $s-color-link;
  font-family: inherit;
  font-size: t-modular-scale(0);
  text-align: center;
  text-decoration: none;
  cursor: pointer;
  transition: color 0.2s, background 0.2s, border-color 0.2s,
    transform 0.1s ease-out;

  @include t-wekkit-hdpi-small-dimension;

  &:hover,
  &:focus {
    color: $s-color-white;
    background-color: $s-color-link;
    border-color: $s-color-link;
    text-decoration: none;
  }

  &:active {
    color: $s-color-white;
    background-color: $s-color-link-active;
    border-color: $s-color-link-active;
    transform: translateY(2px);
  }

  &:focus {
    outline: none;
  }

  /** Primary **/
  &--primary {
    color: $s-color-white;
    background-color: $s-color-link;
    border-color: $s-color-link;

    &:hover,
    &:focus {
      color: $s-color-link;
      background-color: transparent;
      border-color: $s-color-link;
    }

    &:active {
      color: $s-color-link-active;
      background-color: transparent;
      border-color: $s-color-link-active;
    }
  }

  /** Secondary **/
  &--secondary {
    color: $s-color-text-dimmed;
    border-color: $s-color-text-dimmed;

    &:hover,
    &:focus {
      color: $s-color-white;
      background-color: $s-color-text-dimmed;
      border-color: $s-color-text-dimmed;
    }

    &:active {
      color: $s-color-white;
      background-color: darken($s-color-text-dimmed, 10%);
      border-color: darken($s-color-text-dimmed, 10%);
    }
  }

  /** Disabled **/
  &[disabled] {
    cursor: default;
    pointer-events: none;
    user-select: none;
    color: $s-color-gray-darker;
    background-color: $s-color-gray-lighter;
    border-color: currentColor;

    &:active {
      transform: none;
    }
  }

  /** Small **/
  &--sm {
    @include btn-padding(0.5);
  }

  /** Large **/
  &--lg {
    @include btn-padding(1.5);
  }

  /** Full width **/
  &--block {
    width: 100%;
  }

  @media (max-width: t-breakpoint(sm)) {
    &--block-\@sm {
      width: 100%;
    }
  }

  @media (max-width: t-breakpoint(md)) {
    &--block-\@md {
      width: 100%;
    }
  }

  /* Hover spacer */
  > #{$root}__spacer {
    display: inline-block;
    width: 0.5em;
    transition: width 0.2s ease-out;
  }

  &:hover > #{$root}__spacer {
    width: 1em;
  }

  /** Social **/
  &--social {
    *:not(:first-child) {
      margin-left: 0.5em;
    }
  }

  /*** Icon ***/
  &__icon {
    svg {
      height: t-modular-scale(1);
      width: t-modular-scale(1);
      vertical-align: text-top;
    }

    path {
      fill: currentColor;
    }
  }

  /*** Text ***/
  // &__text {}

  /*** Counter ***/
  &__counter {
    font-weight: 500;

    &:empty {
      display: none;
    }
  }

  /** Badge **/
  &--social-badge {
    line-height: 1;
    padding: t-vertical-rhythm(0.4);
  }

  /** Branded color **/
  @mixin btn-social-color($s-color-brand) {
    color: $s-color-brand;
    border-color: $s-color-brand;

    &:hover,
    &:focus {
      color: $s-color-white;
      background: $s-color-brand;
      border-color: $s-color-brand;
    }

    &:active {
      color: $s-color-white;
      background: darken($s-color-brand, 10%);
      border-color: darken($s-color-brand, 10%);
    }
  }

  // Twitter
  &--social-twitter {
    @include btn-social-color($s-color-twitter);
  }

  // Facebook
  &--social-facebook {
    @include btn-social-color($s-color-facebook);
  }

  // Google+
  &--social-googleplus {
    @include btn-social-color($s-color-googleplus);
  }

  // LinkedIn
  &--social-linkedin {
    @include btn-social-color($s-color-linkedin);
  }

  // Pinterest
  &--social-pinterest {
    @include btn-social-color($s-color-pinterest);
  }
}
