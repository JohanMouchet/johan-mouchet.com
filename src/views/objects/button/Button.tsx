import * as React from "react";
import cx from "classnames";
import "./Button.scss";

type Props = {
  type: "anchor" | "button" | "submit" | "reset";
  url?: string;
  variant?: "primary" | "secondary" | "linkedIn";
  size?: "sm" | "lg";
  block?: boolean | "@sm" | "@md";
  className?: string | string[] | { [key: string]: boolean };
  attributes?: boolean | number | string;
  children?: React.ReactNode;
};

const Button: React.FC<Props> = ({
  type = "button",
  url,
  variant,
  size,
  block,
  className,
  children,
  ...attributes
}) => {
  const classNames = cx(
    "o-button",
    size && `o-button--${size}`,
    variant && `o-button--${variant}`,
    typeof block === "string"
      ? `o-button--block${"-" + block}`
      : block === true && "o-button--block",
    className
  );

  return type === "anchor" ? (
    <a className={classNames} href={url} {...attributes}>
      {children}
    </a>
  ) : (
    <button className={classNames} type={type} {...attributes}>
      {children}
    </button>
  );
};

export default Button;
