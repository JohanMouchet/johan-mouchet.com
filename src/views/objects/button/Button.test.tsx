import React from "react";
import { shallow } from "enzyme";
import Button from ".";

it("Renders <Button /> without crashing", () => {
  expect(
    shallow(
      <Button type="anchor" url="#" variant="primary" size="sm" block="@sm">
        children
      </Button>
    )
  ).toMatchSnapshot();
});
