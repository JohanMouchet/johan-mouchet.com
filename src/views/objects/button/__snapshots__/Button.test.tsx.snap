// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders <Button /> without crashing 1`] = `
<a
  class="o-button o-button--sm o-button--primary o-button--block-@sm"
  href="#"
>
  children
</a>
`;
